### DISCUSSION TITLE
Edge count consistently underrepresented in dense network

### DISCUSSION THREAD

**Discussion by dylan-munson** [statnet/ergm#589] [Q&A]
Hello,

I am trying to fit an ERGM to some network data that I have collected (I may be able to share a deidentified version of the data upon request).  The raw network is fairly dense, with 344 nodes and 727 edges, but also with a good amount of isolates and a fairly skewed degree distribution (undirected, range = 0-35).  There is also a very high number of closed triangles in the data (the triad census gives 555 300-type triangles).  The ERGM I am trying to fit has some basic demographic main and homophily effects as well as a GWESP term.

However, the networks that are simulated during the fitting procedure consistently, even at their maximum, have lower edge counts than the observed network.  I also tried simulating a network myself and the closed triangle count was also much lower than in the observed data.  I have tried toying with the decay parameter in the GWESP term to no avail (raising it too high leads to the opposite problem of model degeneracy, in fact).  I have also tried adding terms for isolates and geometrically weighted degree distribution to see if this helps, but in these cases the model usually does not converge or the addition does not help either.  I am somewhat at a loss as I haven't been able to find much about people encountering this kind of problem before.  Any advice would be greatly appreciated and, as I said, I may be able to provide some of the data if anyone would like to have a further look.

EDIT: I forgot to mention that this network also has a fair number of isolates because of the way that the network data was collected.  One reason I am thinking I may be running into this issue is that there are two forces pulling the estimation in opposite directions that end up "meeting in the middle" in an unsatisfactory way--the large number of zero-degree nodes on the one hand and the large number of closed triads amongst the higher-degree nodes on the other.  Short of just removing the isolates, is there perhaps some way I can try to balance out these forces more appropriately?  I have also added a term in the model now to control for the degree of each node (a nodecov("degree") term).  It may also be worth noting that the model seems to have convergence issues in general, with the iterations beginning very far away from the tolerance region in absolute terms and often straying further away again even as the iterations continue.

**Comment by CarterButts** [comment]
You may need to consider some of the component/connectivity statistics from `ergm.components` for this type of network.  However, your description implies that your real problem may be dodgy data.  For instance, you imply that many of your isolates may be spurious.  If you have e.g. spurious isolates because you have converted missing data to null dyads, then you are unlikely to ever get a model that will work well: the reason is because you are trying to build a single ERGM that describes not only the process that generates the actual network, but also that describes whatever data collection/encoding process converted that underlying true network into whatever you actually fit.  Moreover, any model that did "work" would be extremely misleading, since it would not be describing the generator of the actual network, but of that process mixed in some way with your measurement/encoding process.  It's hard to say more from your description, but it sounds like you need to deal with your data problems before proceeding further.

**Comment by dylan-munson** [comment]
@CarterButts Thanks very much for your suggestions.  It probably is worthwhile for me to go back and check the data again in any case.  I guess I should clarify that I actually do believe the isolates are meaningful and theoretically important, in the sense that during data collection, they were people named by respondents as contacts but who were not listed as being people from whom the respondents received information.  Thus I leave them in as "potential" connections in the network who happen to not be connected to the respondent.

That said, I do think it is possible that in the main connected component of the graph, there is some kind of upward bias in the number of edges because of the way the data was collected.  If anything this would be a bigger concern for me than the isolates (especially because removing the isolates seems to have no effect on estimation).  If this is the case, I'm not sure how I could deal with this type of situation or test for it, but if you have any suggestions I would much appreciate hearing them.

**Reply by CarterButts** [reply]
I think it is important to understand how this data was collected.  Your comment implies that (1) edges were obtained by surveying egos and getting tie self-reports, and that (2) some nodes were included that were not actually surveyed.  If that is correct, then you would need to correct for the missing data associated with those nodes: at minimum, all edge variables connecting pairs of those nodes are missing data (`NA`).  It sounds like you are collecting self-reports of incoming edges; if so, you also are missing the incoming edge variables from the surveyed respondents to the non-surveyed nodes.  (If you asked about outgoing contacts, then it's the reverse.  If you asked about both in and out-ties, then only the ties among non-surveyed nodes are missing.)  If you have missing data and coded it as nulls, that will definitely cause problems (not just with model fit, but with interpretation): to anthropomorphize, your model would be desperately trying to find out why those nodes have no ties, and in particular, why they aren't tied to each other.  But in fact, that's a data collection and coding artifact, so even if the model did find a way to fit the data, the solution it would provide would be misleading.  The good news is that `ergm` can handle missing edge data in a smart way (thanks to work by Mark and by Krista Gile), provided that the design used to collect the data was ignorable, and provided that the data is correctly coded as missing.  Similarly, if you have hard constraints in the data collection (e.g., no one was allowed to name more than `k` alters), it can be necessary to condition on these in order for the model to work properly.  Very complex designs can require a lot of correction to account for these things, so it is important to get them right - this is a known (but underappreciated) source of persistent problems with model fit.  (To take a dramatic example, let's say that you have a directed network where everyone was forced to nominate exactly 4 alters.  You will never get the outdegree distribution right - nor anything that depends on it - with a model based on substantive social mechanisms, because no ordinary social mechanism would produce a relationship that everyone would send exactly 4 of.   The poor model then gets the blame...but the problem is that the model is being asked to "explain" something that isn't actually a property of the social process.  However, if you use the `constraints` argument to tell the the model that (by fiat) everyone has outdegree 4, the model now "knows" that this is a fixed property (and not an emergent one), and will be able to do a much better job.)  I would say to take a look at your design, make sure that all missingness and/or constraints are passed to the model (missingness via `NA` coding of edge variables, constraints by the `constraints` argument), and then see if that helps.

**Reply by dylan-munson** [reply]
Thanks for your detailed reply.  I am not sure I entirely understand what you mean by NA coding of edge variables--if you mean edge covariates I'm not actually using any of those, but if you mean edges between non-surveyed respondents then really the whole point of my design is to simulate the full network that includes such edges using the coefficients fit with an ERGM/tapered ERGM.  If there is additional coding that needs to be done to correctly capture that, though, as you seem to suggest, I'd be interested in further references to how to perform such a procedure.
